core/context.go

```diff
@@ -0,0 +1,130 @@
+// limitations under the License.
+
+package core
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/dolthub/dolt/go/libraries/doltcore/doltdb"
+	"github.com/dolthub/dolt/go/libraries/doltcore/sqle/dsess"
+	"github.com/dolthub/go-mysql-server/sql"
+
+	"github.com/dolthub/doltgresql/core/sequences"
+)
+
+// contextValues contains a set of objects that will be passed alongside the context.
+type contextValues struct {
+	collection *sequences.Collection
+}
+
+// contextValuesKey is the key to access the contextValues in the context.
+const contextValuesKey = `@_?dg` // Arbitrary but unique
+
+// getContextValues accesses the contextValues in the given context. If the context does not have a contextValues, then
+// it creates one and adds it to the context.
+func getContextValues(ctx *sql.Context) (*contextValues, error) {
+	valuesAny := ctx.Value(contextValuesKey)
+	if valuesAny == nil {
+		cv := &contextValues{}
+		ctx.Context = context.WithValue(ctx.Context, contextValuesKey, cv)
+		return cv, nil
+	}
+	cv, ok := valuesAny.(*contextValues)
+	if !ok {
+		return nil, fmt.Errorf("context contains an unknown values struct of type: %T", valuesAny)
+	}
+	return cv, nil
+}
+
+// getRootFromContext returns the working session's root from the context, along with the session.
+func getRootFromContext(ctx *sql.Context) (*dsess.DoltSession, *RootValue, error) {
+	session := dsess.DSessFromSess(ctx.Session)
+	// Does this handle the current schema as well?
+	state, ok, err := session.LookupDbState(ctx, ctx.GetCurrentDatabase())
+	if err != nil {
+		return nil, nil, err
+	}
+	if !ok {
+		return nil, nil, fmt.Errorf("cannot find the database while fetching root from context")
+	}
+	return session, state.WorkingRoot().(*RootValue), nil
+}
+
+// GetTableFromContext returns the table from the context. Returns nil if no table was found.
+func GetTableFromContext(ctx *sql.Context, schema string, tableName string) (*doltdb.Table, error) {
+	_, root, err := getRootFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+	table, _, err := root.GetTable(ctx, doltdb.TableName{
+		Name:   tableName,
+		Schema: schema,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return table, err
+}
+
+// GetCollectionFromContext returns the given sequence collection from the context. Returns a nil sequence if it could
+// not be found.
+func GetCollectionFromContext(ctx *sql.Context) (*sequences.Collection, error) {
+	cv, err := getContextValues(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if cv.collection == nil {
+		_, root, err := getRootFromContext(ctx)
+		if err != nil {
+			return nil, err
+		}
+		cv.collection, err = root.GetSequences(ctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return cv.collection, nil
+}
+
+// CloseContextRootFinalizer finalizes any changes persisted within the context by writing them to the working root.
+// This should ONLY be called by the ContextRootFinalizer node.
+func CloseContextRootFinalizer(ctx *sql.Context) error {
```

Comment:
Is this piece of machinery necessary?

If you just set the working set directly in the session during mutation, does that work? That's kind of how everything else works.

This seems to just add another layer of pseudo-transactionality, which we should avoid if we can get away with it. It's the job of transaction code to atomically commit things or not, they're free to change within the session as much as they want.